tests/unit/lms/views/api/blackboard/files_test.py

```diff
@@ -19,14 +24,99 @@ def test_it(
+                        "nextPage": "PAGE_2_PATH"
+                    }
+                }
+            ),
+            # The second response contains the path to the third page.
+            factories.requests.Response(
+                json_data={"paging": {"nextPage": "PAGE_3_PATH"}}
+            ),
+            # The third response is the last page: no path to the next page.
+            factories.requests.Response(json_data={}),
+        ]
+        blackboard_list_files_schema.parse.side_effect = [
+            # The first response from the Blackboard API contains the first
+            # three files.
+            [sentinel.file_1, sentinel.file_2, sentinel.file_3],
+            # The second page's files are in the second response.
+            [sentinel.file_4, sentinel.file_5, sentinel.file_6],
+            # The final page of files.
+            [sentinel.file_7, sentinel.file_8],
+        ]
+
+        files = view()
+
+        # It called the Blackboard API three times getting the three pages.
+        assert blackboard_api_client.request.call_args_list == [
+            call("GET", "courses/uuid:COURSE_ID/resources?limit=200"),
+            call("GET", "PAGE_2_PATH"),
+            call("GET", "PAGE_3_PATH"),
+        ]
+        # It returned all three pages of files as a single list.
+        assert files == [
+            sentinel.file_1,
+            sentinel.file_2,
+            sentinel.file_3,
+            sentinel.file_4,
+            sentinel.file_5,
+            sentinel.file_6,
+            sentinel.file_7,
+            sentinel.file_8,
+        ]
+
+    def test_it_doesnt_send_paginated_requests_forever(
+        self, view, blackboard_api_client, blackboard_list_files_schema
+    ):
+        # Make the Blackboard API send next page paths forever.
+        blackboard_api_client.request.return_value = factories.requests.Response(
+            json_data={"paging": {"nextPage": "NEXT_PAGE"}}
+        )
+
+        files = view()
+
+        assert blackboard_api_client.request.call_count == PAGINATION_MAX_REQUESTS
+        assert len(files) == PAGINATION_MAX_REQUESTS * len(
+            blackboard_list_files_schema.parse.return_value
+        )
+
+    @pytest.fixture(autouse=True)
+    def blackboard_api_client(self, blackboard_api_client):
+        blackboard_api_client.request.return_value = factories.requests.Response(
+            json_data={}
+        )
+        return blackboard_api_client
+
+    @pytest.fixture(autouse=True)
+    def blackboard_list_files_schema(self, blackboard_list_files_schema):
+        blackboard_list_files_schema.parse.return_value = [
+            sentinel.file_1,
+            sentinel.file_2,
+            sentinel.file_3,
+        ]
+        return blackboard_list_files_schema
+
+    @pytest.fixture(autouse=True)
     def pyramid_request(self, pyramid_request):
         pyramid_request.matchdict["course_id"] = "COURSE_ID"
         return pyramid_request
```

Comment:
You can't see it here, but I think every fixture at the bottom of the apart from `views` only applies to one of the test classes or another, and could probably be local to that test class.

Reply:
The `pyramid_request()` one is also needed by all three tests. I localized all the others